frontend/src/components/Layout.js

```diff
@@ -55,86 +58,88 @@ const useStyles = makeStyles(() => ({
                 }
+            });
 
-                return block
-            }
-
-            response.data.forEach(({ upstream, downstream }) => {
-                const upstreamBlock = registerBlock(upstream)
-                const downstreamBlock = registerBlock(downstream, true)
+            edges.forEach(({ upstream, downstream, edge_type }) => {
+                const upstreamBlock = blocks[upstream]
+                const downstreamBlock = blocks[downstream]
                 if (upstreamBlock && downstreamBlock) {
-                    upstreamBlock.downstream[downstreamBlock.ip] = downstreamBlock
-                    downstreamBlock.upstream[upstreamBlock.ip] = upstreamBlock
+                    downstreamBlock.upstream.push({ block: upstreamBlock, edgeType: edge_type })
+                    upstreamBlock.downstream.push({ block: downstreamBlock, edgeType: edge_type })
                 }
-            });
-
-            function repositionDownstream(currentBlock) {
-                let lastYVal = currentBlock.location[1]
-                Object.values(currentBlock.downstream).forEach((block, index) => {
-                    const [cx, cy] = currentBlock.location
-                    block.location = [cx + 55, cy + 50 * index]
-                    lastYVal = Math.max(lastYVal, repositionDownstream(block))
-                })
+            })
 
-                return lastYVal
+            let currentY = 0
+
+            const roots = Object.values(blocks).filter(item => item.upstream.length === 0)
+            const positionBlock = (x, block, currentY) => {
+                let startFromY = currentY
+                if (!block.visited) {
+                    block.location = [x, startFromY]
+                    block.downstream.forEach(downstreamBlock => {
+                        positionBlock(x + 45, downstreamBlock.block, startFromY)
+                        startFromY += 30
+                    })
+                    block.visited = true
+                }
+                return startFromY
             }
 
-            let lastYVal = 30
-            localBlocks.forEach(block => {
-                if (Object.keys(block.upstream).length === 0) {
-                    block.location[1] = lastYVal
-                    lastYVal = repositionDownstream(block)
-                }
+            roots.forEach(root => {
+                currentY = positionBlock(0, root, currentY)
             })
+            setBlocks(Object.values(blocks))
 
-            setBlocks(localBlocks)
-
-            console.log(localBlocks)
-
-            return setGraph(response.data)
+            console.log(blocks)
+            return dispatch(getGraphReducer(edges))
         }
-        fetchData()
+        fetchData();
+    }, [ip, storedIP])
 
-    }, [])
 
     const handleClick = block => () => {
-        setTargetIP(block.ip)
-        setIP(block.ip);
-        window.location.reload();
+        const ip = dispatch(setIPReducer(block.ip))
```

Comment:
this will automatically trigger a change in the store that will backpropagate to this screen.
`ip` and `setIp` can be removed (I think)